src/badguy/ghoul.cpp

```diff
@@ -0,0 +1,213 @@
+//  SuperTux
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#include "badguy/ghoul.hpp"
+
+#include "object/player.hpp"
+#include "editor/editor.hpp"
+#include "sprite/sprite.hpp"
+#include "supertux/game_session.hpp"
+#include "supertux/sector.hpp"
+#include "util/reader_mapping.hpp"
+#include "util/writer.hpp"
+
+static const float FLYSPEED = 80.0f; /**< speed in px per second */
+static const float TRACK_RANGE = 2500.0f; /**< at what distance to start tracking the player */
+
+Ghoul::Ghoul(const ReaderMapping& reader) :
+  BadGuy(reader, "images/creatures/ghoul/ghoul.sprite"),
+  PathObject(),
+  m_mystate(STATE_IDLE),
+  m_flyspeed(),
+  m_track_range()
+{
+  reader.get("flyspeed", m_flyspeed, FLYSPEED);
+  reader.get("track-range", m_track_range, TRACK_RANGE);
+  
+  bool running;
+  if ( !reader.get("running", running)) running = false;
+
+  init_path(reader, running);
+  
+  m_sprite->set_action(m_dir == Direction::LEFT ? "left" : "right", /* loops = */ -1);
+}
+
+bool
+Ghoul::collision_squished(GameObject& object)
+{
+  auto player = Sector::get().get_nearest_player(m_col.m_bbox);
+  if (player)
+    player->bounce (*this);
+  m_sprite->set_action("squished", 1);
+  kill_fall();
+  return true;
+}
+
+bool
+Ghoul::is_freezable() const
+{
+  return false;
+}
+
+bool
+Ghoul::is_flammable() const
+{
+  return false;
+}
+
+void
+Ghoul::finish_construction()
+{
+  if (get_walker() && get_walker()->is_running()) {
+    m_mystate = STATE_PATHMOVING_TRACK;
+  }
+}
+
+void
+Ghoul::activate()
+{
+  if (Editor::is_active())
+    return;
+}
+
+void
+Ghoul::deactivate()
+{
+  switch (m_mystate) {

```

Comment:
I don't think you have to list all the states here.

You can simply just list the state that does something and at the end do
```c++
default:
  break;
```

Reply:
fixed

Reply:
I think that you could also use a simple if statement to make it even more simple.
```C++
if (m_mystate == STATE_TRACKING)
  mystate = STATE_IDLE;
```